### ISSUE TITLE
test_c10d.py: running with -pt option causes failing tests to pass

### ISSUE THREAD

**Issue by jfc4050** [pytorch/pytorch#25491]
## 🐛 Bug
running tests with `-pt` option for `test_c10d.py` allows failing tests to pass. I tried to reproduce this in some other test files, including `test_distributed.py` and it doesn't seem to be an issue there.

## To Reproduce
add something like this to `test_c10d.py`
```python
    def test_demonstrate_pytest_problem(self):
        10 / 0
```
run `➜ python test/run_test.py -v -pt -i c10d` and get `test_c10d.py::ProcessGroupGlooTest::test_demonstrate_pytest_problem PASSED                                                                                                          [ 30%]`

now run `➜ python test/run_test.py -v -i c10d`and get as expected:
```
Traceback (most recent call last):
  File "/home/justin/anaconda3/envs/torch/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/justin/anaconda3/envs/torch/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/justin/pytorch/test/common_distributed.py", line 140, in _run
    getattr(self, self.id().split(".")[2])()
  File "/home/justin/pytorch/test/common_distributed.py", line 103, in wrapper
    fn(self)
  File "test_c10d.py", line 743, in test_demonstrate_pytest_problem
    10 / 0
```

## Expected behavior

`test_demostrate_pytest_problem` should fail for both pytest and unittest runs

## Environment
PyTorch version: 1.3.0a0+d2a8435
Is debug build: No
CUDA used to build PyTorch: Could not collect

OS: Ubuntu 19.04
GCC version: (Ubuntu 8.3.0-6ubuntu1) 8.3.0
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: No
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 418.56
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.3.0a0+763b35b
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.0.2            py37h7b6447c_0  
[conda] mkl_fft                   1.0.14           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] torch                     1.3.0a0+d2a8435           dev_0    <develop>

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera

**Comment by yf225** [comment]
@mrshenli @pietern Would you like to take a look?

**Comment by jfc4050** [comment]
oh and also
```bash
➜ pip freeze | grep pytest
pytest==5.0.1
```

**Comment by zhaojuanmao** [comment]
it was being fixed in #25441
